File: example/modskel/modskel.c
#define AUGAS_SHARED
#define AUGAS_BUILD
#include "augas/module.h"

#include <stdlib.h>

static const char* modname_ = NULL;
static const struct augas_host* host_ = NULL;

static void
close_(const struct augas_session* s)
{
    host_->writelog_(modname_, AUGAS_LOGINFO, "close_()");
}

static int
open_(struct augas_session* s, const char* serv, const char* peer)
{
    host_->writelog_(modname_, AUGAS_LOGINFO, "open_()");
    return 0;
}

static int
data_(const struct augas_session* s, const char* buf, size_t size)
{
    host_->writelog_(modname_, AUGAS_LOGINFO, "data_()");
    host_->send_(s->sid_, buf, size, AUGAS_SESSELF);
    return 0;
}

static int
rdexpire_(const struct augas_session* s, unsigned* ms)
{
    host_->writelog_(modname_, AUGAS_LOGINFO, "rdexpire_()");
    return 0;
}

static int
wrexpire_(const struct augas_session* s, unsigned* ms)
{
    host_->writelog_(modname_, AUGAS_LOGINFO, "wrexpire_()");
    return 0;
}

static int
stop_(const struct augas_session* s)
{
    host_->writelog_(modname_, AUGAS_LOGINFO, "stop_()");
    host_->shutdown_(s->sid_);
    return 0;
}

static int
event_(int type, void* arg)
{
    host_->writelog_(modname_, AUGAS_LOGINFO, "event_()");
    return 0;
}

static int
expire_(void* arg, unsigned id, unsigned* ms)
{
    host_->writelog_(modname_, AUGAS_LOGINFO, "expire_()");
    return 0;
}

static int
reconf_(void)
{
    host_->writelog_(modname_, AUGAS_LOGINFO, "reconf_()");
    return 0;
}

static const struct augas_module fntable_ = {
    close_,
    open_,
    data_,
    rdexpire_,
    wrexpire_,
    stop_,
    event_,
    expire_,
    reconf_
};

static const struct augas_module*
load_(const char* modname, const struct augas_host* host)
{
    modname_ = modname;
    host_ = host;
    host_->writelog_(modname_, AUGAS_LOGINFO, "load_()");
    return &fntable_;
}

static void
unload_(void)
{
    host_->writelog_(modname_, AUGAS_LOGINFO, "unload_()");
    host_ = 0;
}

AUGAS_MODULE(load_, unload_)

File: example/module/modskel.py
import augas

# void reconf (void);
# void writelog (int level, string msg);
# void post (string sname, int type, object user);
# string getenv (string sname, string name);
# int tcpconnect (string sname, string host, string serv);
# int tcplisten (string sname, string host, string serv);
# int settimer (string sname, int tid, unsigned ms, object user);
# bool resettimer (string sname, int tid, unsigned ms);
# bool canceltimer (string sname, int tid);
# void shutdown (int fid);
# void send (string sname, int cid, buffer buf);
# void setrwtimer (int cid, unsigned ms, unsigned flags);
# void resetrwtimer (int cid, unsigned ms, unsigned flags);
# void cancelrwtimer (int cid, unsigned flags);

class Buffer:
    def __init__(self):
        self.tail = ''

    def lines(self, data):
        data = self.tail + data
        ls = data.split('\n')
        self.tail = ls.pop()
        for l in ls:
            yield l

def logdebug(s):
    augas.writelog(augas.LOGDEBUG, s)

def closesess(sname):
    logdebug("closesess(): %s" % sname)

def opensess(sname):
    augas.writelog(augas.LOGDEBUG, "opensess(): %s" % sname)
    augas.tcplisten(sname, "0.0.0.0", "8080")

def event(sname, type, user):
    logdebug("event(): %s" % sname)

def expire(sname, tid, user, ms):
    logdebug("expire(): %s" % sname)

def reconf(sname):
    logdebug("reconf(): %s" % sname)

def closeconn(sname, cid, user):
    logdebug("closeconn(): %s" % sname)

def openconn(sname, cid, addr, port):
    logdebug("openconn(): %s" % sname)
    augas.setrwtimer(cid, 5000, augas.TIMRD)
    return Buffer()

def notconn(sname, cid, user):
    logdebug("notconn(): %s" % sname)

def data(sname, cid, user, buf):
    logdebug("data(): %s" % sname)
    for line in user.lines(str(buf)):
        augas.send(sname, cid, line + '\n', augas.SNDSELF)

def rdexpire(sname, cid, user, ms):
    logdebug("rdexpire(): %s" % sname)
    augas.send(sname, cid, 'are you there?\n', augas.SNDSELF)

def wrexpire(sname, cid, user, ms):
    logdebug("wrexpire(): %s" % sname)

def teardown(sname, cid, user):
    logdebug("teardown(): %s" % sname)
    for line in user.lines(str(buf)):
        augas.send(sname, cid, user.tail + '\n', augas.SNDSELF)
    augas.shutdown(cid)
